### ISSUE TITLE
Small models don't trampoline (was: Opaque requirements for MH)

### ISSUE THREAD

**Issue by longouyang** [probmods/webppl#51]
```js
MH(function() {
  return flip(0.9)
}, 4000);
```
fails with an error:

```
/Users/long/webppl/src/header.js:841
  var fw = -Math.log(oldTrace.length);
           ^
RangeError: Maximum call stack size exceeded
    at Array.map (native)
    at mhAcceptProb (/Users/long/webppl/src/header.js:842:26)
    at MH.exit (/Users/long/webppl/src/header.js:856:22)
    at exit (/Users/long/webppl/src/header.js:493:13)
    at MH.sample (/Users/long/webppl/src/header.js:824:3)
    at MH.exit (/Users/long/webppl/src/header.js:881:15)
    at exit (/Users/long/webppl/src/header.js:493:13)
    at MH.sample (/Users/long/webppl/src/header.js:824:3)
    at MH.exit (/Users/long/webppl/src/header.js:881:15)
    at exit (/Users/long/webppl/src/header.js:493:13)
```

(I am on the latest dev version, 8e645405009fb3a395e5a035c290c16d886d9abc)

But this modified version where we name the ERP works:
```js
var numSamps = 4000;
MH(function() {
  var x = flip(0.9);
  return x;
}, numSamps);
```

```
* Program return value:

ERP:
    true: 0.903
    false: 0.097
```

Why does writing it the first way result in a stack overflow?

**Comment by stuhlmueller** [comment]
Trampolining (which makes unrestricted tail recursion work in Javascript) only applies to the model, not the header. If the model is extremely small, then trampolining doesn't have anything to work with and there is no point in the recursion induced by the MH implementation where we clear the stack. This is another argument for exploring a global tail-call transform using the 6to5 transpiler (issue #45), which would presumably also apply to the header.

**Comment by ngoodman** [comment]
Would it be enough to wrap the model in a dummy function, to force trampolining? If so, that would be simple and not too ugly...

**Comment by stuhlmueller** [comment]
Yeah, a wrapping that does what Long's modified version does would help. Transform...

    [model defs]

into

    var result = function(){
      [model defs]
      return ...
    }()
    return result;

**Comment by null-a** [comment]
A little more detail about what's happening here:

Really tiny models *can* work in some situations. This is fine for example:

```
PMCMC(function () {
  return 1;
}, 10, 5000);
```

The compiled model code includes a single closure which is enough to have trampolining work in this case.[1]

The stack overflow problem arises specifically when using models of one variable with MH. When written in the original compact form, compilation introduces zero closures between the `flip` and the end of the model. Since MH will always regenerate from the `flip`, no closures are encountered during inference, and as @stuhlmueller said trampolining has nothing to work with.

Just about any other way you can think of writing the model *will* introduce additional closures, fixing the problem.

This suggests an alternate (though not necessarily better) solution. We can check (at runtime) whether the continuation used for regeneration is the `exit` continuation, and in such cases insert a closure to ensure trampolining occurs. It looks like [this](https://github.com/null-a/webppl/commit/c908c9808c5787ef54a2e91fa5c7b53a50fe2f8e).

(We'd also need to add this PFRj, but it would hopefully only be in one place once we've refactored.)

[1] I simplified this a bit. PMCMC [introduces an extra closure](https://github.com/probmods/webppl/blob/dev/src/inference/pmcmc.js#L73) which would make this work even without the closure in the model. However, without the closure introduced by PMCMC, the tiny model still works because of the closure in the model itself.

**Comment by stuhlmueller** [comment]
> This suggests an alternate (though not necessarily better) solution. We can check (at runtime) whether the continuation used for regeneration is the exit continuation, and in such cases insert a closure to ensure trampolining occurs. It looks like this.

I think it would be slightly better to do the wrapping as part of the transformation process. Otherwise, we'll have to remember to do it for any MH-like inference algorithm, and do it in multiple places. Doing it as part of the transform could also potentially only do the wrapping if it's necessary, and not do it for models that are 'big enough' as is.

**Comment by null-a** [comment]
> I think it would be slightly better to do the wrapping as part of the transformation process.

If I understand the suggestion correctly then in the transform we'd need to identify the functions which are passed to an inference algorithm (as the `wpplFn` argument) then wrap them if they're too 'small'. This doesn't appear to be straight-forward in general, as you might have this kind of thing:

```
var model = function() { ... };
var Infer = Enumerate;
Infer(model);
```

Am I missing something?

**Comment by stuhlmueller** [comment]
Yes, you wouldn't be able to detect this in all cases, only when the argument is directly a function expression, i.e. `Infer(function(){...})`; if the argument is something else (a symbol, application, etc), you'd want to do the wrapping just to be sure, even if it may not be necessary.

Since a single extra closure probably doesn't make much of a performance difference, you could ignore this complication and always transform `Infer(model)` to something like `Infer(function(){var x = model(); return x;})`.

**Comment by null-a** [comment]
My experimental refactoring in #64 side steps this problem. Since we have a work-around, I'll hold off implementing a specific fix in the hope that this goes away as we clean up the inference algorithms.

**Comment by null-a** [comment]
Closed by #169.
